### ISSUE TITLE
Specify OS version

### ISSUE THREAD

**Issue by sigrlami** [haskell/cabal#2517]
Is there some way to extend my cabal file and specify which *version* of os is currently in use and load specific modules?

AFAIK, Cabal's `conditional` doesn't provide such mechanism, but something like

```haskell
if os(osx) && ver(=10.10)
   hs-source-dirs: osx10
if os(osx) && ver(<=10.9)
   hs-source-dirs: osx9
```
will be nice, where `ver` is keyword for version of os

If there is no such thing, will it be acceptable if I submit patch for it?

**Comment by 23Skidoo** [comment]
Can't you work around this with CPP and maybe a `configure` script? Otherwise I think something like this could be accepted, but the concept of a single OS version doesn't really apply to all OSs.

/cc @dcoutts

**Comment by sigrlami** [comment]
I'm trying to reduce use of CPP as much as possible, with help of technique described here http://blog.haskell-exists.com/yuras/posts/stop-abusing-cpp-in-haskell.html with something like
```Haskell
extra-source-files: linux/System/..
                    osx/System/..
hs-source-dirs: src
if os(linux)
  hs-source-dirs: linux
if os(darwin)
  hs-source-dirs: osx
```
where extra sources contains specific implementation for platform.

Just now I have only mac and linux users, but might be Windows in a couple of months, and on Windows I might have same issue, with some changes between Windows7 and Windows10. So I want to handle it now, without need of bringing additional build mechanism.

I want to make useful contribution, but can keep it to my fork, if it's outside cabal's architecture.

**Comment by andrewthad** [comment]
This seems useful. Any thoughts from the maintainers on whether or not such a feature is considered within the scope of cabal?

**Comment by angerman** [comment]
I wonder if we could make the `os` an implicity comparable.

```
ios(osx == 10.10) 
```
`osx` would be `False` on non-macOS platforms, and otherwise defined to be the target version, which unless compared is considered `True`?

**Comment by hvr** [comment]
@angerman there's precedent for that: `impl(ghc [<constraint>])` works like that; if we go for this, we should mimic its semantics

**Comment by phadej** [comment]
OS versions are a bit tricky in general. 

What’s linux version? Kernel’s, Ubuntu/Debian/... release, coretools?

Same OSX release may run different darwin kernels: https://en.m.wikipedia.org/wiki/Darwin_(operating_system)

Windows?

So, what’s this check about: kernel, some userland thing?

**Comment by 23Skidoo** [comment]
Agreed, I'd like to see a more detailed proposal for the semantics of this conditional, at least for the three main supported platforms (OS X, Windows, and Linux). I still have the feeling that this use case is better handled by `configure` scripts which allow more granular tests.

**Comment by andrewthad** [comment]
@phadej That's a good question. What I had in mind was just the kernel version, both for linux and osx. That is, not supporting constraints like Ubuntu >= 18.04 or OSX >= 10.8. I don't know how the windows kernel is versioned.

@23Skidoo What kind of `configure` script are you suggesting? Doing something in `Setup.hs` or using `autoconf`?

**Comment by 23Skidoo** [comment]
@andrewthad `build-type: Configure` is what I had in mind, but `build-type: Custom` can also work. For the Windows case, see https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions

**Comment by hvr** [comment]
There's also https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/cross_development/Using/using.html#//apple_ref/doc/uid/20002000-SW7 if you want to introspect OSX versions.

However, both `configure` and `Custom` build-types have the problem that they act long after the solver has already computed a build-plan. So once the configure/custom Setup.hs are invoked you need to cope with whatever version of the OS you have as can't backtrack to a different build-plan from there anymore, nor can you conditionally depend on different dependencies based on the OS version.

**Comment by MuhammedZakir** [comment]
I would also like this. For macOS, I would like to check against macOS version. I don't think there is a kernel ver. vs OS ver. issue for macOS as OS version is bumped when kernel is updated. The wikipedia article linked above shows this. I think this is same for Windows. For other OSes, I am not sure. If it's hard to make a decision on that, is it possible to add support for at least macOS and Windows?

**Comment by yaitskov** [comment]
I suggest to try to use TemplateHaskell, because it runs at build time and can run IO. 
Run check in IO action and pick appropriate TH slice. Must work if imports don't change for any mac os version, which are enabled all at once through `os_darwin`.
